### PR TITLE
38 display donor

### LINKED ISSUE #38: Implement Feature to Display Donor Details on Website
Develop a new feature that allows for the display of donor details on the website. This involves modifying the navigation bar to include a link to a new page where these details are listed. The existing `"Donor Form"` button in the navbar should be replaced with a `"Donors"` link. This link will direct users to the newly created `/donors` path, where a detailed list of all donors can be viewed. Additionally, relocate the `"Donor Form"` to be accessible from the `/donors` page, ensuring it is still easily accessible to users who need to submit new donor information.

Image

**Tasks:**

- [x] Update the Navbar:

    - Replace the current "Donor Form" link in the navbar with a new "Donors" link.
    - Ensure the "Donors" link routes to the newly created /donors page.

- [x] Create the Donors Page:

    - Implement a new page at /donors that will display a list of all donor details.
    - Design the page to be user-friendly and ensure it aligns with the website’s existing aesthetic.
    - Include functionality to fetch and display donor details from the backend efficiently.

- [x] Relocate the Donor Form:

    - Move the "Donor Form" to be accessible via the /donors page.
    - Ensure the form is fully functional in its new location and continues to handle submissions properly.

- [x] Frontend Routing Adjustments:

    - Update frontend routing to reflect these changes and ensure seamless navigation.

- [x] Testing:

    - Conduct thorough testing to ensure the new page and updated navbar work correctly across different devices and browsers.
    - Test the functionality of fetching and displaying the donor details to verify performance and accuracy.

Note:  As we are gradually transitioning to TypeScript, please ensure that any new files created for this feature are written in TypeScript. This is part of our ongoing effort to enhance our codebase's scalability and maintainability.

### PR DESCRIPTION
**Pull request by truffer11**
# Pull Request Template

**Fixes #38 **

### What was changed?

- Adds a new donor page where you can see a list of all donors
- DonatedItemsList.js was updated to .tsx
- Enums.js was updated to .tsx
- Added "All Donors" section under Donors dropdown in navbar

### Why was it changed?

- The .js files were updated to .tsx to add additional security, as part of our ongoing project of updating old js code to tsx
- The donor page allows for administration to find donors easily and get details on each donor

### How was it changed?

- Utilized the .tsx code from DonatedItemsList to ensure we kept the same look across the website
- Currently, there are five columns for the donors: Donor ID, First Name, Last Name, email, and more details. More details is where you will be able to get more specifics including address and stuff. This was done to make the list more readable, while still giving enough information to verify you are looking at the right donor.

### Screenshots that show the changes (if applicable):

- **After:**
  - Screenshot 2024-10-28 145622

### CODE REVIEW

File: client-app/src/Components/DonatedItemsList.tsx
Comment (miabu-pashh) [inline client-app/src/Components/DonatedItemsList.tsx:1]: I have worked on this file change, you can revert the files changes in this file.

File: client-app/src/Components/Navbar.tsx [hunk 1]
```diff
@@ -120,7 +120,7 @@ const Navbar: React.FC = () => {
                                     <ul className="dropdown-menu" aria-labelledby="navbarDropdown">
                                         <li>
                                             {/*Input at a later time, once we have an all donors page. Please also add the bold to the fontWeight and color above */}
-                                            {/*<Link className="dropdown-item" to="/donations">All Donors</Link> */}
+                                            <Link className="dropdown-item" to="/donorlist">All Donors</Link>
```
Comment (miabu-pashh) [inline client-app/src/Components/Navbar.tsx:123]: reduce the comments and try to edit the existing link.

File: client-app/src/constants/Enums.ts
Comment (miabu-pashh) [inline client-app/src/constants/Enums.ts:1]: I guess Enums file change is not necessary for your issue.

File: client-app/src/Components/DonorList.tsx
Comment (miabu-pashh) [inline client-app/src/Components/DonorList.tsx:1]: the table is touching the top navbar.  create a css file for this list component and update the styles.

**Review by miabu-pashh — COMMENTED** [review]
1. Some files needs to be reverted, as it was done by other team members for the same team sprint.
2. need to add css
3. minimize the comments lines
4. run the prettier check before the pull request.

File: client-app/src/Components/DonorList.tsx [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+                            <p>Donor Last Name: {selectedItemDetails.lastName}</p>
+                            <p>Donor Email: {selectedItemDetails.email}</p>
+                        </div>
+                    )}
+                    <button onClick={() => setModalIsOpen(false)}>Close</button>
+                </Modal>
+                <div
+                    style={{ position: 'fixed', bottom: '20px', right: '20px' }}
+                >
+                    <button onClick={() => handleAddNewDonationClick()}>
```
Comment (Anjali0407-git) [inline client-app/src/Components/DonorList.tsx:192]: Function name should be relevant to Donors, please change it

File: client-app/src/Components/DonorList.tsx [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+                                <td>{item.firstName}</td>
+                                <td>{item.lastName}</td>
+                                <td>{item.email}</td>
+                                <td>
+                                    <button onClick={() => handleViewDetailsClick()}>
+                                        View More Details
+                                     </button>
+                                </td>
+
+                                {assignProgramClicked && (
```
Comment (Anjali0407-git) [inline client-app/src/Components/DonorList.tsx:156]: I see that you added `assignProgramClicked` in couple of places in this page. What is this for?

**Review by Anjali0407-git — CHANGES_REQUESTED** [review]

**Comment by Anjali0407-git** [comment]
@truffer11 Is this ready for re-review?

File: client-app/src/Components/DonorList.tsx [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    const [currentDonors, setCurrentDonors] = useState<Donor[]>([]); // initially empty array
+    const [error, setError] = useState<string | null>(null);
+    const navigate = useNavigate();
+
+    useEffect(() => {
+        // Fetch donor data from the backend API
+        const fetchDonors = async () => {
+            try {
+                const response = await axios.get<Donor[]>( // This will likely need to be updated, but this is the only way I could get this to work
+                    //`${process.env.REACT_APP_BACKEND_API_BASE_URL}/donor`
```
Comment (Anjali0407-git) [inline client-app/src/Components/DonorList.tsx:37]: @truffer11 Please replace the hardcoded data with environment variables

File: client-app/src/Components/DonorList.tsx [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+                                <button onClick={() => handleViewDetailsClick(donor)}>
+                                    View More Details
+                                </button>
+                            </td>
+                        </tr>
+                    ))}
+                </tbody>
+            </table>
+
+            <Modal isOpen={modalIsOpen} onRequestClose={() => setModalIsOpen(false)}>
```
Comment (Anjali0407-git) [inline client-app/src/Components/DonorList.tsx:130]: @truffer11 Please remove this modal. We do not need donor details in program page

File: client-app/src/Components/DonorListTest.jsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
+import React from 'react';
+import { render, screen, fireEvent, waitFor } from '@testing-library/react';
+import axios from 'axios';
+import DonorList from './DonorList'; // Adjust the import based on your file structure
+
+jest.mock('axios');
+
+describe('DonorList Component', () => {
```
Comment (Anjali0407-git) [inline client-app/src/Components/DonorListTest.jsx:8]: @truffer11 we have seperate tests folder for writing tests, can you please move this test suite to there? also ensure you follow the same file naming convention .

Comment (Anjali0407-git) [reply client-app/src/Components/DonorListTest.jsx:8]: Also there are merge conflicts

File: server/src/views/error.jade [hunk 7]
```diff
@@ -0,0 +1,7 @@
+doctype html
+html
+  head
+    title Error
```
Comment (Anjali0407-git) [inline server/src/views/error.jade:4]: @truffer11 remove this file please

**Comment by Anjali0407-git** [comment]
All changes looks god, but merge conflicts??...

**Pull request merged**